tests/Application/Actions/Hooks/StripeUpdateTest.php

```diff
@@ -0,0 +1,179 @@
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+
+        $request = $this->createRequest('POST', '/hooks/stripe', $body)
+            ->withHeader('Stripe-Signature', $this->generateSignature($time, $body, $webhookSecret));
+        
+        $response = $app->handle($request);
+
+        $this->assertEquals(204, $response->getStatusCode());
+    }
+
+    public function testMissingSignature(): void
+    {
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        $body = file_get_contents(dirname(__DIR__, 3) . '/TestData/success.json');
+
+        $donationRepoProphecy = $this->prophesize(DonationRepository::class);
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $stripeRepoProphecy = $this->prophesize(StripeWebhook::class);
+
+        $container->set(StripeWebhook::class, $stripeRepoProphecy->reveal());
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+
+        $request = $this->createRequest('POST', '/hooks/stripe', $body)
+            ->withHeader('Stripe-Signature', '');
+        
+        $response = $app->handle($request);
+
+        $expectedPayload = new ActionPayload(400, ['error' => [
+            'type' => 'BAD_REQUEST',
+            'description' => 'Invalid Signature',
+        ]]);
+        $expectedSerialised = json_encode($expectedPayload, JSON_PRETTY_PRINT);
+
+        $payload = (string) $response->getBody();
+
+        $this->assertEquals($expectedSerialised, $payload);
+        $this->assertEquals(400, $response->getStatusCode());
+    }
+
+    public function testSuccess(): void
+    {
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        $body = file_get_contents(dirname(__DIR__, 3) . '/TestData/success.json');
+        $donation = $this->getTestDonation();
+        $webhookSecret = (string) $container->get('settings')['stripe']['webhookSecret'];
+        $time = (string) time();
+
+        $donationRepoProphecy = $this->prophesize(DonationRepository::class);
+        $donationRepoProphecy
+            ->findOneBy(['transactionId' => 'pi_externalId_123'])
+            ->willReturn($donation)
+            ->shouldBeCalledOnce();
+
+        $donationRepoProphecy
+            ->push(Argument::type(Donation::class), false)
+            ->willReturn(true)
+            ->shouldBeCalledOnce();
+
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $stripeRepoProphecy = $this->prophesize(StripeWebhook::class);
+
+        $container->set(StripeWebhook::class, $stripeRepoProphecy->reveal());
```

Comment:
It's interesting that this works, because:

1. the container is not the source of `StripeWebhook` – you are instantiating it directly inside `action()` instead of getting it through dependency injection
2. the prophecy has no methods set up, so if it was being used it would never return the test data or verify a signature

If everything is being covered this suggests that maybe we didn't need to mock Stripe's SDK at all to cover this endpoint, in which case we don't need these container overrides and prophecies, and we quite possibly don't need the concrete class I suggested either (sorry!)

Reply:
Thanks for the clarification @NoelLH think i'm getting this hang of this prophecy-wizardry business.

It looks like we still need to concrete class, doing the following, throws `Error: Cannot instantiate abstract class Stripe\\Webhook`

```
use Stripe\Webhook;

...

$webhook = new Webhook();
```

Reply:
@tbg-jasonfung you cannot instantiate the abstract class, but that's not what we were doing before. Does calling the method statically as before work?